src/unity/toolkits/activity_classification/activity_classifier.cpp

```diff
@@ -628,45 +671,70 @@ void activity_classifier::perform_training_iteration() {
   size_t num_classes = read_state<size_t>("num_classes");
   size_t prediction_window = read_state<size_t>("prediction_window");
   
-  while (training_data_iterator_->has_next_batch()) {
-    data_iterator::batch batch =
-        training_data_iterator_->next_batch(batch_size);
 
-    // Submit the batch to the neural net model.
-    // TODO: Implement double buffering.
-    std::map<std::string, shared_float_array> results = training_model_->train(
-        { { "input",   batch.features },
-          { "labels",  batch.labels   },
-          { "weights", batch.weights  }, });
-    const shared_float_array &loss_batch = results.at("loss");
+  // To support double buffering, use a queue of pending inference results.
+  std::queue<result> pending_batches;
 
-    float batch_loss = std::accumulate(loss_batch.data(),
-                                       loss_batch.data() + loss_batch.size(),
-                                       0.f, std::plus<float>());
+  auto pop_until_size = [&](size_t remaining) {
 
-    
-    cumulative_batch_loss += batch_loss / batch.batch_info.size();
+    while (pending_batches.size() > remaining) {
 
-    
-    const shared_float_array& output = results.at("output");
+      // Pop one batch from the queue.
+      result batch = pending_batches.front();
+      pending_batches.pop();
 
-    cumulative_batch_accuracy +=
-        cumulative_chunk_accuracy(prediction_window, num_classes, output,
-                                  batch) /
-        batch.batch_info.size();
+      cumulative_batch_accuracy += cumulative_chunk_accuracy(prediction_window, num_classes, batch.output_info,
+                                    batch.data_info ) / batch.data_info.batch_info.size();
 
-    ++num_batches;
+      float batch_loss = std::accumulate(batch.loss_info.data(),
+                                         batch.loss_info.data() + batch.loss_info.size(),
+                                         0.f, std::plus<float>());
+
+      cumulative_batch_loss += batch_loss / batch.data_info.batch_info.size();
+
+    }
+  };
+
+  // Iterate through the data once.
+  data_iterator::batch input_batch =
```

Comment:
Let's follow the pattern of the original code and move the call to `next_batch` inside the loop... (as you had it for the validation-data case)